finality-aleph/src/testing/mocks/validator_network.rs

```diff
@@ -90,3 +112,373 @@ impl<D: Data> MockNetwork<D> {
+    }
+}
+
+pub struct UnreliableConnectionMaker {
+    c_dialers: mpsc::UnboundedReceiver<(Address, oneshot::Sender<Connection>)>,
+    c_listeners: Vec<mpsc::UnboundedSender<Connection>>,
+}
+
+impl UnreliableConnectionMaker {
+    pub fn new(ids: Vec<AuthorityId>) -> (Self, Callers, Addresses) {
+        let mut c_listeners = Vec::with_capacity(ids.len());
+        let mut callers = HashMap::with_capacity(ids.len());
+        let (tx_dialer, c_dialers) = mpsc::unbounded();
+        // create peer addresses that will be understood by the main loop in method run
+        // each peer gets a one-element vector containing its index, so we'll be able
+        // to retrieve proper communication channels
+        let addr: Addresses = ids
+            .clone()
+            .into_iter()
+            .zip(0..ids.len())
+            .map(|(id, u)| (id, vec![u as u32]))
+            .collect();
+        // create callers for every peer, keep channels for communicating with them
+        for id in ids.into_iter() {
+            let (tx_listener, rx_listener) = mpsc::unbounded();
+            let dialer = MockDialer {
+                c_connect: tx_dialer.clone(),
+            };
+            let listener = MockListener {
+                c_new_connection: rx_listener,
+            };
+            c_listeners.push(tx_listener);
+            callers.insert(id, (dialer, listener));
+        }
+        (
+            UnreliableConnectionMaker {
+                c_dialers,
+                c_listeners,
+            },
+            callers,
+            addr,
+        )
+    }
+
+    pub async fn run(&mut self, connections_end_after: usize) {
+        loop {
+            info!(target: "validator-network", "UnreliableConnectionMaker: waiting for new request...");
+            let (addr, c) = self.c_dialers.next().await.expect("should receive");
+            info!(target: "validator-network", "UnreliableConnectionMaker: received request");
+            let (l_stream, r_stream) = Connection::new(4096, connections_end_after);
+            info!(target: "validator-network", "UnreliableConnectionMaker: sending stream");
+            c.send(l_stream).expect("should send");
+            self.c_listeners[addr as usize]
+                .unbounded_send(r_stream)
+                .expect("should send");
+        }
+    }
+}
+
+#[derive(Clone)]
+struct MockData {
+    data: u32,
+    filler: Vec<u8>,
+    decodes: bool,
+}
+
+impl MockData {
+    fn new(data: u32, filler_size: usize, decodes: bool) -> MockData {
+        MockData {
+            data,
+            filler: vec![0; filler_size],
+            decodes,
+        }
+    }
+}
+
+impl Encode for MockData {
+    fn encode(&self) -> Vec<u8> {
+        // currently this is exactly the default behaviour of codec
+        let mut encoded = Vec::with_capacity(self.filler.len() + 5);
+        encoded.append(&mut u32::encode(&self.data));
+        encoded.append(&mut Vec::<u8>::encode(&self.filler));
```

Comment:
oh, and why is that?

Reply:
`self.filler` should have `size_hint` implemented. It might be better to use that. Same for `u32` and `bool`.

Reply:
The canonical implementation of `encode` in the `Codec` trait for `Vec::<u8>` is most likely: "write 4 bytes for size + write the bytes". This is because `Codec` works on streams not on vectors, so in paritcular `decode` gets a `stream` on input, not a vector.

Reply:
A kind of related note -- for efficiency one should always implement `using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R` of the `Encode` trait instead of `encode`. This does not matter too much (especially here, in tests), but it typically allows to avoid some allocations.

Reply:
In general, I heavily recommend the readme here https://github.com/paritytech/parity-scale-codec  and actually reading the `Encode/Decode` trait definitions, and default implementations therein. Lots to learn.

Reply:
oh, sorry, I totally misunderstood what you're talking about - I thought that you mean that **everything** will take +4 bytes, not the `Vec`... then of course, you're right

Reply:
fixed